pkgs/native_assets_cli/lib/src/api/build_output.dart

```diff
@@ -4,48 +4,113 @@
+  /// Start time for the build of this output.
   ///
-  /// Rounded down to whole seconds, because [File.lastModified] is rounded
-  /// to whole seconds and caching logic compares these timestamps.
+  /// The [timestamp] is rounded down to whole seconds, because
+  /// [File.lastModified] is rounded to whole seconds and caching logic compares
+  /// these timestamps.
   DateTime get timestamp;
-  List<Asset> get assets;
-  Dependencies get dependencies;
-  Metadata get metadata;
 
+  /// The assets produced by this build.
+  ///
+  /// In dry runs, the assets for all [Architecture]s for the [OS] specified in
+  /// the dry run must be provided.
+  Iterable<Asset> get assets;
+
+  /// The files used by this build.
+  ///
+  /// If any of the files in [dependencies] are modified after [timestamp], the
+  /// build will be re-run.
+  Iterable<Uri> get dependencies;
+
+  /// Create a build output.
+  ///
+  /// The [timestamp] must be before any [dependencies] are read by the build
+  /// this output belongs to. If the [BuildOutput] object is created at the
+  /// beginning of the `build.dart` script, [timestamp] can be omitted and will
+  /// default to [DateTime.now]. The [timestamp] is rounded down to whole
+  /// seconds, because [File.lastModified] is rounded to whole seconds and
+  /// caching logic compares these timestamps.
+  ///
+  /// The [Asset]s produced by this build or dry-run can be provided to the
+  /// constructor as [assets], or can be added later using [addAsset] and
+  /// [addAssets]. In dry runs, the [Architecture] for [CCodeAsset]s can be
+  /// omitted.
+  ///
+  /// The files used by this build must be provided to the constructor as
+  /// [dependencies], or can be added later with [addDependency] and
+  /// [addDependencies]. If any of these files are modified after [timestamp],
+  /// the build will be re-run. Typically these dependencies contain the
+  /// `build.dart` script itself, and the source files used in the build.
+  ///
+  /// Metadata can be passed to `build.dart` invocations of dependent packages.
+  /// It must be provided to the constructor as [metadata], or added later with
+  /// [addMetadatum] and [addMetadata].
   factory BuildOutput({
     DateTime? timestamp,
-    List<Asset>? assets,
-    Dependencies? dependencies,
-    Metadata? metadata,
+    Iterable<Asset>? assets,
+    Iterable<Uri>? dependencies,
+    Map<String, Object>? metadata,
   }) =>
-      model.BuildOutput(
+      BuildOutputImpl(
         timestamp: timestamp,
-        assets: assets?.map((e) => e as model.Asset).toList(),
-        dependencies: dependencies as model.Dependencies?,
-        metadata: metadata as model.Metadata?,
+        assets: assets?.cast<AssetImpl>().toList(),
+        dependencies: Dependencies([...?dependencies]),
+        metadata: Metadata({...?metadata}),
       );
 
-  /// The version of [BuildOutput].
+  /// Adds [Asset]s produced by this build or dry run.
+  void addAsset(Asset asset);
```

Comment:
ditto: This concept of `asset` is specific to native assets (it has `linkMode` / ...).

Imagine we want to extend this for flutter builds to include
* `config.java.{jdk,...}` and `config.android.{minSdk,...}`
* `output.java.jars`

There's  no `linkMode` needed, not even asset id.

Reply:
There is no `linkMode` in `DataAsset`.

`Asset` only has `id` and `file`.

Whether `id` should be optional is a good question.
If asset-id is only used for platform-independent-access at runtime in Dart code, and a Jar is linked in so it isn't accessed from Dart. Probably for static libraries a similar argument can be made, e.g. they will also be linked in.
I wonder if we should expand the concept of `asset id` to be an identifier that's useful for other purposes as well. Let's say linking the jar or static lib fails, would it be useful to give an error message with the id? Maybe not, we already have a unique file name.

As for extensibility, I'm happy to introduce a `NonPredefinedAsset` that has a `String type` and `Map info` which would give you arbitrary JSON access.

(As for making the whole thing JSON. I think I don't want to try that rewrite in this PR. I'd rather do that in a follow up PR.)